core/ipc/grpc/client/src/main/java/org/opennms/core/ipc/grpc/client/MinionGrpcClient.java

```diff
@@ -0,0 +1,509 @@
+            LOG.info("Removing module {} from gRPC IPC client.", rpcModule.getId());
+        }
+    }
+
+    private boolean hasChangedToReadyState() {
+        ConnectivityState prevState = currentChannelState;
+        return !prevState.equals(ConnectivityState.READY) && getChannelState().equals(ConnectivityState.READY);
+    }
+
+    public void shutdown() {
+        requestHandlerExecutor.shutdownNow();
+        blockingSinkMessageScheduler.shutdownNow();
+        registerdModules.clear();
+        if (rpcStream != null) {
+            rpcStream.onCompleted();
+        }
+        channel.shutdown();
+        LOG.info("Minion at location {} with systemId {} stopped", minionIdentity.getLocation(), minionIdentity.getId());
+    }
+
+    public void setBundleContext(BundleContext bundleContext) {
+        this.bundleContext = bundleContext;
+    }
+
+
+    @Override
+    public String getMetricDomain() {
+        return SINK_METRIC_PRODUCER_DOMAIN;
+    }
+
+    @Override
+    public BundleContext getBundleContext() {
+        return bundleContext;
+    }
+
+    @Override
+    public Tracer getTracer() {
+        if (tracerRegistry != null) {
+            return tracerRegistry.getTracer();
+        }
+        return GlobalTracer.get();
+    }
+
+    @Override
+    public MetricRegistry getMetrics() {
+        if (metrics == null) {
+            return new MetricRegistry();
+        }
+        return metrics;
+    }
+
+    public void setMetrics(MetricRegistry metrics) {
+        this.metrics = metrics;
+    }
+
+    public TracerRegistry getTracerRegistry() {
+        return tracerRegistry;
+    }
+
+    public void setTracerRegistry(TracerRegistry tracerRegistry) {
+        this.tracerRegistry = tracerRegistry;
+    }
+
+    ConnectivityState getChannelState() {
+        return currentChannelState = channel.getState(true);
+    }
+
+
+    @Override
+    public <S extends Message, T extends Message> void dispatch(SinkModule<S, T> module, String metadata, T message) {
+
+        try (Logging.MDCCloseable mdc = Logging.withPrefixCloseable(MessageConsumerManager.LOG_PREFIX)) {
+            byte[] sinkMessageContent = module.marshal(message);
+            String messageId = UUID.randomUUID().toString();
+            SinkMessage.Builder sinkMessageBuilder = SinkMessage.newBuilder()
+                    .setMessageId(messageId)
+                    .setLocation(minionIdentity.getLocation())
+                    .setModuleId(module.getId())
+                    .setContent(ByteString.copyFrom(sinkMessageContent));
+
+            if (module.getId().equals(HEARTBEAT_MODULE_ID)) {
+                if (rpcStream == null || sinkStream == null || hasChangedToReadyState()) {
+                    initializeSinkStub();
+                    initializeRpcStub();
+                }
+            }
+            setTagsForSink(sinkMessageBuilder);
+            // If module has asyncpolicy, keep attempting to send message.
```

Comment:
If module *doesn't* have async policy, then block?

Reply:
If module has async policy set then we should block here till we are able to send this message.
If module has no async policy set, we should attempt to send now and don't block dispatch.